File: js/src/test/timestretch.test.ts
/**
 * WSOLA time-stretch: properties, deliberately not bytes.
 *
 * There is **no shared byte-level fixture** for this feature, in any of the five
 * ports, and that is a decision rather than an omission. The alignment search
 * ranks candidate offsets by a cross-correlation whose last bit depends on the
 * order a language sums floats in; one offset chosen differently moves every
 * sample after it, so a golden file would fail for a reason that is not a
 * defect, and the usual response to a fixture that fails for no reason is to
 * regenerate it — which switches the check off. The four properties asserted
 * here fail only when the behaviour is actually wrong: the output length is
 * exact, the pitch does not move (a resampler would move it by exactly `speed`),
 * the loudness survives, and `speed = 1.0` is a bypass rather than a stretch by
 * one.
 */

import assert from "node:assert/strict";
import { test } from "node:test";

import { MAX_SPEED, MIN_SPEED, stretchedLength, timeStretch } from "../timestretch.js";

const RATE = 24_000;
const SPEEDS = [0.5, 0.8, 1.25, 1.5, 2.0];

/**
 * A voiced-sounding test signal: a fundamental plus its second harmonic, so the
 * alignment search has real periodic structure to lock onto rather than the
 * single lobe of a pure tone.
 */
function tone(f0: number, samples: number): Float32Array {
  const x = new Float32Array(samples);
  for (let i = 0; i < samples; i++) {
    const t = i / RATE;
    x[i] = Math.sin(2 * Math.PI * f0 * t) + 0.5 * Math.sin(4 * Math.PI * f0 * t);
  }
  return x;
}

/**
 * Fundamental by autocorrelation, over the lag range of a human voice.
 *
 * Unnormalised on purpose: the biased sum decays slowly with lag, which breaks
 * the tie between a period and twice a period in favour of the period — the
 * octave error every naive pitch tracker makes. The decay is far too gentle to
 * move the peak within its own lobe.
 */
function fundamental(x: Float32Array): number {
  const minLag = Math.floor(RATE / 500);
  const maxLag = Math.floor(RATE / 80);
  let bestLag = minLag;
  let best = -Infinity;
  for (let lag = minLag; lag <= maxLag; lag++) {
    let sum = 0;
    for (let i = 0; i + lag < x.length; i++) sum += x[i] * x[i + lag];
    if (sum > best) {
      best = sum;
      bestLag = lag;
    }
  }
  return RATE / bestLag;
}

function rms(x: Float32Array): number {
  let sum = 0;
  for (let i = 0; i < x.length; i++) sum += x[i] * x[i];
  return Math.sqrt(sum / Math.max(1, x.length));
}

test("speed 1.0 gives back the very same array", () => {
  // Identity, not equality. The default must not enter the DSP at all: every
  // conformance vector and every existing caller has to be unaffected by this
  // feature existing, and "no arithmetic happened" is easier to trust than "the
  // arithmetic was the identity".
  const x = tone(220, 4_000);
  assert.equal(timeStretch(x, RATE, 1.0), x);
});

test("the output length is exactly floor(n / speed + 0.5)", () => {
  const n = 9_973; // odd, and not a multiple of the frame or the hop
  const x = tone(200, n);
  for (const speed of SPEEDS) {
    const out = timeStretch(x, RATE, speed);
    assert.equal(out.length, stretchedLength(n, speed), `speed ${speed}`);
    assert.equal(out.length, Math.floor(n / speed + 0.5), `speed ${speed}`);
  }
});

test("stretchedLength rounds halves up, the way the other four ports do", () => {
  // 5 / 2 is exactly 2.5. `round()` in Python answers 2 (halves to even) and
  // `Math.round` answers 3, so the literal `floor(x + 0.5)` is what keeps the
  // ports agreeing rather than agreeing-except-on-exact-halves.
  assert.equal(stretchedLength(5, 2.0), 3);
  assert.equal(stretchedLength(3, 2.0), 2);
  assert.equal(stretchedLength(24_000, 1.5), 16_000);
  assert.equal(stretchedLength(0, 1.5), 0);
});

test("pitch does not move and loudness survives", () => {
  const f0 = 220;
  const x = tone(f0, RATE / 2);
  const before = fundamental(x);
  for (const speed of [0.5, 1.5, 2.0]) {
    const out = timeStretch(x, RATE, speed);
    const after = fundamental(out);
    // A resampler would land at f0 * speed here — 110 Hz or 440 Hz — so 3 % is
    // a wide margin against the failure this is guarding, and a tight one
    // against a stretch that has started smearing.
    assert.ok(
      Math.abs(after - before) / before < 0.03,
      `speed ${speed}: pitch moved ${before.toFixed(1)} Hz -> ${after.toFixed(1)} Hz`
    );
    const ratio = rms(out) / rms(x);
    assert.ok(
      Math.abs(ratio - 1) < 0.15,
      `speed ${speed}: RMS ratio ${ratio.toFixed(3)} is outside 15 %`
    );
    assert.ok(
      out.every((v) => Number.isFinite(v)),
      `speed ${speed}: the output is not finite everywhere`
    );
  }
});

test("a speed outside the offered range is refused, not clamped", () => {
  const x = tone(220, 4_000);
  for (const bad of [0.49, 0, -1, 2.01, 3.0]) {
    assert.throws(
      () => timeStretch(x, RATE, bad),
      new RegExp(`outside \\[${MIN_SPEED}, ${MAX_SPEED}\\]`),
      `speed ${bad} was not refused`
    );
  }
  // Non-finite is its own message: NaN is neither inside nor outside a range,
  // and reporting it as "outside [0.5, 2.0]" would send the caller looking at
  // their bounds rather than at where their NaN came from.
  for (const bad of [NaN, Infinity, -Infinity]) {
    assert.throws(() => timeStretch(x, RATE, bad), /finite/, `speed ${bad} was not refused`);
  }
  // The bounds themselves are inside the range.
  assert.equal(timeStretch(x, RATE, MIN_SPEED).length, stretchedLength(x.length, MIN_SPEED));
  assert.equal(timeStretch(x, RATE, MAX_SPEED).length, stretchedLength(x.length, MAX_SPEED));
});

test("a fragment shorter than one frame is cut or padded, not stretched", () => {
  // Below one frame there is no second frame to align against, so the correct
  // answer is the right *length* filled with what there is — wrong in the way
  // silence is wrong rather than in the way a pitch shift is.
  const x = tone(220, 100);
  const faster = timeStretch(x, RATE, 2.0);
  assert.equal(faster.length, 50);
  assert.deepEqual(Array.from(faster), Array.from(x.subarray(0, 50)));

  const slower = timeStretch(x, RATE, 0.5);
  assert.equal(slower.length, 200);
  assert.deepEqual(Array.from(slower.subarray(0, 100)), Array.from(x));
  assert.ok(
    slower.subarray(100).every((v) => v === 0),
    "the pad is not silence"
  );

  // An empty input stays empty rather than throwing on a zero-length window.
  assert.equal(timeStretch(new Float32Array(0), RATE, 1.5).length, 0);
});

test("the frame is derived from the sample rate, not hardcoded", () => {
  // The same fragment is above one frame at 8 kHz (200 samples) and below it at
  // 48 kHz (1200), so a port that hardcodes 600 samples takes the wrong branch
  // at one of these two rates and the difference is a whole feature silently not
  // running.
  //
  // The signal has to be non-stationary for the two branches to be
  // distinguishable at all: a perfectly steady tone compressed 1.5x *is* its own
  // prefix, sample for sample, because nothing in it happens at a particular
  // time. An amplitude ramp gives the stretch something to move, and makes the
  // difference between the branches legible — a compressed ramp reaches its top,
  // a truncated one stops two thirds of the way up.
  const n = 1_200;
  const x = tone(220, n);
  for (let i = 0; i < n; i++) x[i] *= i / n;
  const peak = (a: Float32Array) => a.reduce((m, v) => Math.max(m, Math.abs(v)), 0);

  const at8k = timeStretch(x, 8_000, 1.5);
  const at48k = timeStretch(x, 48_000, 1.5);
  assert.equal(at8k.length, stretchedLength(n, 1.5));
  assert.equal(at48k.length, stretchedLength(n, 1.5));

  // 48 kHz: one frame is 1200 samples, so this fragment is not longer than a
  // frame and the output is the input's head, cut.
  assert.deepEqual(Array.from(at48k), Array.from(x.subarray(0, at48k.length)));

  // 8 kHz: one frame is 200 samples, so the overlap-add ran instead. Same input,
  // same speed, same output length, different samples — and the ramp climbs
  // further, because a compressed ramp keeps rising where a truncated one simply
  // stops.
  assert.notDeepEqual(Array.from(at8k), Array.from(at48k));
  assert.ok(peak(at8k) > peak(at48k), "the overlap-add path only truncated the ramp");
});

test("a sample rate too low to have a hop does not hang", () => {
  // The guard that no implementation tested, which is how two of the five —
  // this one included — shipped without it.
  //
  // Below ~60 Hz the derived frame is one sample, so the hop (frame / 2) is
  // zero, and `writeAt += hop` never advances. This port computed the hop
  // *after* the degenerate-shape guard, so `timeStretch(new Float32Array(64),
  // 40, 1.5)` ran forever while Python returned 43 samples in microseconds.
  // Nothing was red, because nothing asked.
  //
  // Bounded by the clock as well as by the length: a regression here is a hang,
  // and a hang wedges CI rather than failing it.
  const started = Date.now();
  const got = timeStretch(new Float32Array(64), 40, 1.5);
  assert.ok(Date.now() - started < 5_000, "the overlap-add loop did not terminate");
  assert.equal(got.length, stretchedLength(64, 1.5));
  assert.equal(got.length, 43);
});

test("two calls agree bit for bit", () => {
  // No RNG, no adaptivity, no wall clock. Same in, same out, forever — which is
  // what lets the engine promise the same bytes for the same seed *and* the
  // same speed.
  const x = tone(220, RATE);
  const first = timeStretch(x, RATE, 1.4);
  const second = timeStretch(x, RATE, 1.4);
  assert.deepEqual(Array.from(first), Array.from(second));
});

test("silence stays silent", () => {
  // The one case that exercises the `energy <= 0` branch of the alignment
  // search: every candidate frame has zero energy, so the normalised
  // cross-correlation would divide by nothing. A missing guard there shows up
  // as NaN across the whole output rather than as a wrong sample.
  const got = timeStretch(new Float32Array(RATE), RATE, 1.5);
  assert.equal(got.length, stretchedLength(RATE, 1.5));
  for (let i = 0; i < got.length; i++) {
    assert.equal(got[i], 0, `silence came back as ${got[i]} at sample ${i}`);
  }
});

test("nothing clips or goes non-finite", () => {
  // Overlap-add divides by an accumulated window; a denominator that reaches
  // zero where the numerator does not is an infinity, and one infinity in a
  // waveform is a click loud enough to hurt.
  const x = tone(220, RATE);
  for (const speed of SPEEDS) {
    const got = timeStretch(x, RATE, speed);
    for (let i = 0; i < got.length; i++) {
      assert.ok(Number.isFinite(got[i]), `speed ${speed} produced ${got[i]} at sample ${i}`);
      assert.ok(
        Math.abs(got[i]) <= 1.2 * 1.5,
        `speed ${speed} produced ${got[i]} at sample ${i}, well past the input's range`
      );
    }
  }
});

File: go/timestretch/timestretch_test.go
// WSOLA: the properties, not the samples.
//
// Nothing here pins a waveform, and there is deliberately no shared byte-level
// fixture across the five ports. The alignment search ranks candidates by a
// cross-correlation whose last bit can differ between languages — Go sums a
// frame of products in a plain loop, NumPy sums it pairwise — and one offset
// chosen differently moves every sample after it. A fixture would then fail for
// a reason that is not a defect, which is the worst kind of red test: the one
// people learn to re-record.
//
// So what is asserted is what a listener would notice if this broke: the
// length, the pitch, the loudness, and the fact that speed 1.0 is not a stretch
// of factor one but a bypass. Python, Rust, TypeScript and Swift assert the same
// properties.

package timestretch

import (
	"math"
	"testing"
	"time"
)

const sampleRate = 24_000

var speeds = []float64{0.5, 0.75, 0.9, 1.25, 1.5, 2.0}

// signal is a voiced-ish test signal: a low fundamental, a harmonic, and a
// sweep. Deterministic by construction — no RNG anywhere in this file, because
// the stretcher has none either and a flaky DSP test is worse than no DSP test.
func signal(seconds, f0 float64) []float32 {
	n := int(float64(sampleRate) * seconds)
	out := make([]float32, n)
	for i := range out {
		t := float64(i) / sampleRate
		out[i] = float32(0.5*math.Sin(2*math.Pi*f0*t) +
			0.25*math.Sin(2*math.Pi*2*f0*t) +
			0.15*math.Sin(2*math.Pi*600*t*(1+t)))
	}
	return out
}

// pitchHz estimates the fundamental by autocorrelation. Enough to catch a
// chipmunk, which is the failure this is looking for — a resampler would move
// the fundamental by exactly speed.
func pitchHz(x []float32) float64 {
	const window = 4096
	at := sampleRate / 4
	w := make([]float64, window)
	mean := 0.0
	for i := range w {
		w[i] = float64(x[at+i])
		mean += w[i]
	}
	mean /= window
	for i := range w {
		w[i] -= mean
	}
	lo, hi := sampleRate/500, sampleRate/80
	bestLag, best := lo, math.Inf(-1)
	for lag := lo; lag < hi; lag++ {
		sum := 0.0
		for i := 0; i+lag < window; i++ {
			sum += w[i] * w[i+lag]
		}
		if sum > best {
			best, bestLag = sum, lag
		}
	}
	return sampleRate / float64(bestLag)
}

func rms(x []float32) float64 {
	sum := 0.0
	for _, s := range x {
		sum += float64(s) * float64(s)
	}
	return math.Sqrt(sum / float64(len(x)))
}

// stretch is TimeStretch with the error turned into a failure, because every
// call below passes a speed the range already allows — the refusal has its own
// test, and repeating four lines of error handling around each property would
// bury the property.
func stretch(t *testing.T, audio []float32, rate int, speed float64) []float32 {
	t.Helper()
	got, err := TimeStretch(audio, rate, speed)
	if err != nil {
		t.Fatal(err)
	}
	return got
}

// Identity, not equality. The engine's default must not depend on a DSP path
// being lossless — it must not enter the DSP path at all.
func TestUnitySpeedHandsBackTheSameSlice(t *testing.T) {
	x := signal(0.3, 220)
	got := stretch(t, x, sampleRate, 1.0)
	if len(got) != len(x) || &got[0] != &x[0] {
		t.Fatal("speed 1.0 returned a new slice; it must be an exact bypass")
	}
}

func TestTheOutputIsExactlyAsLongAsAsked(t *testing.T) {
	x := signal(1.0, 220)
	for _, speed := range speeds {
		got := stretch(t, x, sampleRate, speed)
		if want := StretchedLength(len(x), speed); len(got) != want {
			t.Errorf("speed %v gave %d samples, want %d", speed, len(got), want)
		}
	}
}

// Python rounds halves to even; Go, Rust, Swift and JavaScript do not. A
// one-sample disagreement on an exact half is found six months later, in a
// conformance run, by somebody else — so the formula is written as
// floor(n/speed + 0.5) in all five and never as the language's round().
func TestTheLengthFormulaIsHalfUpNotHalfEven(t *testing.T) {
	if got := StretchedLength(5, 2.0); got != 3 {
		t.Errorf("StretchedLength(5, 2.0) = %d, want 3 (2.5 rounds up, not to even)", got)
	}
}

// The guard that no implementation tested, which is how two of the five shipped
// without it.
//
// Below ~60 Hz the derived frame is one sample, so the hop — frame/2 — is zero,
// and the overlap-add loop advances by it. TypeScript and Swift computed the hop
// *after* the degenerate-shape guard and never tested it, so both looped forever
// on an input Go returned from in microseconds; nothing was red, because nothing
// asked. Now all five ask.
func TestASampleRateTooLowToHaveAHopDoesNotHang(t *testing.T) {
	done := make(chan []float32, 1)
	go func() {
		got, err := TimeStretch(make([]float32, 64), 40, 1.5)
		if err != nil {
			t.Error(err)
		}
		done <- got
	}()
	select {
	case got := <-done:
		if want := StretchedLength(64, 1.5); len(got) != want || want != 43 {
			t.Fatalf("got %d samples, want %d", len(got), want)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("the overlap-add loop did not terminate")
	}
}

// No overlap to align, so it is cut or padded. At 24 kHz this is under 25 ms —
// below anything the engine renders, and the alternative is an index out of
// range on the degenerate case.
func TestAFragmentShorterThanAFrameIsStillTheRightLength(t *testing.T) {
	tiny := make([]float32, 64)
	for i := range tiny {
		tiny[i] = 1
	}
	if got := stretch(t, tiny, sampleRate, 2.0); len(got) != 32 {
		t.Errorf("half of 64 samples came back as %d", len(got))
	}
	got := stretch(t, tiny, sampleRate, 0.5)
	if len(got) != 128 {
		t.Fatalf("double of 64 samples came back as %d", len(got))
	}
	for i := 64; i < 128; i++ {
		if got[i] != 0 {
			t.Fatalf("the pad at %d is %v, want silence", i, got[i])
		}
	}
}

// The entire point of WSOLA rather than a resampler.
func TestPitchIsUnmoved(t *testing.T) {
	x := signal(2.0, 220)
	want := pitchHz(x)
	for _, speed := range speeds {
		got := pitchHz(stretch(t, x, sampleRate, speed))
		if math.Abs(got-want)/want > 0.03 {
			t.Errorf("speed %v moved the fundamental from %.1f Hz to %.1f Hz", speed, want, got)
		}
	}
}

// Overlap-add with a window that does not sum to one is the classic way to get
// a 6 dB drop or a comb filter. The periodic Hann at 50 % overlap sums to one,
// and the denominator corrects the ends.
func TestLoudnessSurvives(t *testing.T) {
	x := signal(1.0, 220)
	want := rms(x)
	for _, speed := range speeds {
		got := rms(stretch(t, x, sampleRate, speed))
		if math.Abs(got-want)/want > 0.15 {
			t.Errorf("speed %v moved RMS from %.4f to %.4f", speed, want, got)
		}
	}
}

func TestNothingClipsOrGoesNonFinite(t *testing.T) {
	x := signal(1.0, 220)
	for _, speed := range speeds {
		for i, s := range stretch(t, x, sampleRate, speed) {
			if math.IsNaN(float64(s)) || math.IsInf(float64(s), 0) {
				t.Fatalf("speed %v produced %v at sample %d", speed, s, i)
			}
			if math.Abs(float64(s)) > 1.2 {
				t.Fatalf("speed %v produced %v at sample %d, well past the input's range", speed, s, i)
			}
		}
	}
}

// The correlation search divides by the candidate's energy; a silent frame is
// where that division has to not happen.
func TestSilenceStaysSilent(t *testing.T) {
	got := stretch(t, make([]float32, sampleRate), sampleRate, 1.5)
	for i, s := range got {
		if s != 0 {
			t.Fatalf("silence came back as %v at sample %d", s, i)
		}
	}
}

// No RNG, no adaptivity, no wall clock. Same in, same out, forever.
func TestTwoCallsAgreeBitForBit(t *testing.T) {
	x := signal(1.0, 220)
	first := stretch(t, x, sampleRate, 1.4)
	second := stretch(t, x, sampleRate, 1.4)
	for i := range first {
		if first[i] != second[i] {
			t.Fatalf("two runs differ at sample %d: %v vs %v", i, first[i], second[i])
		}
	}
}

// A caller who asked for 3x and silently got 2x has a bug only a stopwatch
// finds.
func TestOutOfRangeIsRefusedNotClamped(t *testing.T) {
	for _, speed := range []float64{0.49, 2.01, 0.0, -1.0, 10.0} {
		if err := ValidateSpeed(speed); err == nil {
			t.Errorf("speed %v was accepted", speed)
		}
	}
}

// NaN compares false against both bounds, so a naive range test would let it
// through and produce a waveform of NaNs.
func TestNaNAndInfinityAreRefusedBeforeTheRangeCheck(t *testing.T) {
	for _, speed := range []float64{math.NaN(), math.Inf(1), math.Inf(-1)} {
		if err := ValidateSpeed(speed); err == nil {
			t.Errorf("speed %v was accepted", speed)
		}
	}
}

func TestTheBoundsThemselvesAreAllowed(t *testing.T) {
	for _, speed := range []float64{MinSpeed, 1.0, MaxSpeed} {
		if err := ValidateSpeed(speed); err != nil {
			t.Errorf("speed %v was refused: %v", speed, err)
		}
	}
}

// An out-of-range speed reaching this far is a caller that skipped
// ValidateSpeed, and it is refused rather than silently bypassed: handing back
// the input would be the exact failure MinSpeed refuses by name — asked for 3x,
// silently got 1x, and only a stopwatch finds it. The engine refuses first,
// where the refusal saves six seconds of generation; this is the last line.
func TestAnUnvalidatedSpeedIsRefusedRatherThanBypassed(t *testing.T) {
	x := signal(0.3, 220)
	// TimeStretch directly, not the stretch helper: the helper turns an error
	// into a test failure, which is the wrong shape for the one case where the
	// error is the behaviour under test.
	got, err := TimeStretch(x, sampleRate, 4.0)
	if err == nil {
		t.Fatal("an out-of-range speed was accepted")
	}
	if got != nil {
		t.Fatal("a refusal must not also hand back samples")
	}
}
